### PR TITLE
numpy v1.26.0 with meson

### PR DESCRIPTION
**Pull request by h-vetinari**
Let's see if we can build with meson (esp. on windows...)

### CODE REVIEW

**Comment by h-vetinari** [comment]
I based this off of the scipy recipe, and this went pretty far on unix:
```
Successfully built numpy-1.26.0b1-cp310-cp310-linux_x86_64.whl
+ pip install dist/numpy-1.26.0b1-cp310-cp310-linux_x86_64.whl
Processing ./dist/numpy-1.26.0b1-cp310-cp310-linux_x86_64.whl
Installing collected packages: numpy
Successfully installed numpy-1.26.0b1
```

However we then have:
```
INFO:conda_build.build:Packaging numpy
Packaging numpy
Packaging numpy-1.26.0b1-py310ha4c1d20_0
INFO:conda_build.build:Packaging numpy-1.26.0b1-py310ha4c1d20_0
No files or script found for output numpy
WARNING:conda_build.build:No files or script found for output numpy
number of files: 0
```

I think this might be because the default meson install sets a local `--destdir`:
```
[...]
[505/505] Linking target numpy/random/_generator.cpython-310-x86_64-linux-gnu.so
+ $PREFIX/bin/python $SRC_DIR/vendored-meson/meson/meson.py install --quiet --no-rebuild --destdir $SRC_DIR/.mesonpy-8wjmup75/install
[0/196] numpy/core/lib/libnpymath.a
[1/196] numpy/core/include/numpy/__multiarray_api.c
[...]
```

On windows, the same invocation as on unix fails:
```
(%BUILD_PREFIX%) %SRC_DIR%>%PREFIX%\python.exe -m build -w -n -x     -Cbuilddir=builddir     -Csetup-args=-Dblas=blas     -Csetup-args=-Dlapack=lapack     -Csetup-args=${MESON_ARGS// / -Csetup-args=} 
ERROR Source / does not appear to be a Python project: no pyproject.toml or setup.py
```

This is not a very comprehensive take yet, just a first shot. But it's getting late here, so this'll have to suffice for tonight.

**Comment by rgommers** [comment]
> I think this might be because the default meson install sets a local `--destdir`:

I would not expect that to be relevant, it's an implementation detail that's contained inside what `meson-python` is doing. 

If the produced intermediate wheel is empty, `pip` would not report successfully installing it. If it's not empty, then I'm unclear why `conda-build` doesn't find any files. Perhaps easier to figure out from a local build than a CI log?

**Comment by h-vetinari** [comment]
> If it's not empty, then I'm unclear why `conda-build` doesn't find any files.

My suspicion is that it gets installed somewhere other than `$PREFIX` resp. `SP_DIR`. 

> Perhaps easier to figure out from a local build than a CI log?

Sure. As I said I haven't had much time for this yesterday, just started with opening the PR.

**Comment by h-vetinari** [comment]
So the devdocs about building are not up-to-date, but that's being tracked, and at I also found https://github.com/numpy/numpy/pull/24279. Not sure what's up with pkgconfig, our netlib blas/lapack builds do have the respective `.pc` files.

**Comment by h-vetinari** [comment]
So for cross-compiling osx-arm64, we get
```
Checking for size of "long double" : 8 

../numpy/core/meson.build:461:2: ERROR: Problem encountered: Unknown long double format of size: 8
```
which seems to imply that the [patterns](https://github.com/numpy/numpy/blob/v1.26.0b1/numpy/core/meson.build#L380-L463) used for detecting the `long double` flavour are perhaps not exhaustive.

@matoro, I see you authored https://github.com/numpy/numpy/commit/f1cddd3a159f0fdba8bb744f6e721a5242859474, could it be that osx-arm64 has a different in-memory representation than what you've coded up for size 8?

**Comment by matoro** [comment]
> So for cross-compiling osx-arm64, we get
> 
> ```
> Checking for size of "long double" : 8 
> 
> ../numpy/core/meson.build:461:2: ERROR: Problem encountered: Unknown long double format of size: 8
> ```
> 
> which seems to imply that the [patterns](https://github.com/numpy/numpy/blob/v1.26.0b1/numpy/core/meson.build#L380-L463) used for detecting the `long double` flavour are perhaps not exhaustive.
> 
> @matoro, I see you authored [numpy/numpy@f1cddd3](https://github.com/numpy/numpy/commit/f1cddd3a159f0fdba8bb744f6e721a5242859474), could it be that osx-arm64 has a different in-memory representation than what you've coded up for size 8?

Hey, I just tried to copy the python code 1-for-1.  Does it somehow work on the setuptools implementation?  Could you run this snippet and share the output?

```
#include <stdio.h>

const struct {
    const char before[16];
    const long double x;
    const char after[8];
} foo = {{'\0'}, -123456789.0, {'\0'}};

int main(void)
{
    for (size_t i = 0; i < sizeof(long double); i++) {
        fprintf(stdout, "%04o ", ((const unsigned char*)&foo.x)[i]);
    }
    fprintf(stdout, "\n");
    return 0;
}
```

**Comment by rgommers** [comment]
The size of `long double` is 8 bytes on macOS arm64, and the format is `IEEE_DOUBLE_LE` in a native build. Since you're trying to cross-compile here, probably something is going wrong. The check does:
```meson
meson.get_compiler('c').run(...)
```
It should run under QEMU I think, and it likely doesn't here, or that is not working (can't tell from the CI log). You can backport https://github.com/numpy/numpy/pull/24414 to bypass this check for now.

**Comment by h-vetinari** [comment]
Thanks for the response @matoro and the reference @rgommers. This is looking good now!

Is the successful build with the beta enough from your POV (I'm guessing with a passing test suite we should be good normally?), or do you want to publish it to an rc-label to be able to test it against other packages?

**Comment by rgommers** [comment]
> Is the successful build with the beta enough from your POV (I'm guessing with a passing test suite we should be good normally?), or do you want to publish it to an rc-label to be able to test it against other packages?

This is fine I think for now; quite a few downstream packages are testing in CI themselves, because they already need it for Python 3.12 support. So I don't think we'll learn much more from a testing exercise in conda-forge now.

Thanks for verifying this beta though, great to see it working.

File: recipe/bld.bat [hunk 1]
```diff
@@ -1,23 +1,17 @@
...
+    -Cbuilddir=builddir ^
+    -Csetup-args=-Dblas=blas ^
+    -Csetup-args=-Dlapack=lapack
+if %ERRORLEVEL% neq 0 exit 1
 
-XCOPY %RECIPE_DIR%\f2py.bat %SCRIPTS% /s /e
-if errorlevel 1 exit 1
-
-del %SCRIPTS%\f2py.exe
-if errorlevel 1 exit 1
```
Comment (h-vetinari) [inline recipe/bld.bat:23]: @conda-forge/numpy 

Not sure what the existing setup for f2py on windows was here, resp. what constraints we have to pay attention to. Currently, `meta.yaml` sets a windows-specific entry point, and the artefact (for 1.25.2) contains
```
Scripts/f2py-script.py
Scripts/f2py.bat
Scripts/f2py.exe
```
in `%PREFIX%`.

Calling `f2py -h` still works successfully in this PR, but I'm not sure if we need to pay attention to something else here.

Comment (h-vetinari) [reply recipe/bld.bat:23]: Perhaps @ocefpaf knows (or remembers from https://github.com/conda-forge/numpy-feedstock/pull/96).

Comment (ocefpaf) [reply recipe/bld.bat:23]: I think I copied that from the anaconda recipe. We probably don't need it anymore. I think it was a limitation with how -build handled the scripts.

**Comment by mattip** [comment]
I see the remaining CI failure is PyPy + windows, where `meson --version` [is failing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=777737&view=logs&j=5be07ae1-d8ba-5406-47b6-8e3a3a12f825&t=0bf03e01-0bec-5b85-5316-b1633322e895&l=3086) with an `AttributeError: 'NoneType' object has no attribute 'encoding'`.

**Comment by mattip** [comment]
This seems to be some kind of pytest ineteraction? Using this works with no error:
```
conda create -n pypy3.9 pypy python=3.9
conda activate pypy3.9
mamba install meson
pypy -c "import subprocess; subprocess.check_call(["meson", "--version"])
```

**Comment by h-vetinari** [comment]
> This seems to be some kind of pytest interaction?

And/or meson interaction
```
  File "%PREFIX%\lib\site-packages\mesonbuild\mesonmain.py", line 230, in ensure_stdout_accepts_unicode
    if sys.stdout.encoding and not sys.stdout.encoding.upper().startswith('UTF-'):
AttributeError: 'NoneType' object has no attribute 'encoding'
```

**Comment by h-vetinari** [comment]
Anyway, it looks like this'll have to get fixed on the meson-side. I don't think we should hold up 1.26 of win+pypy for now, it's already in high demand due to the impending 3.12.

**Pull request merged**